### DISCUSSION TITLE
Tcl parser - requiring 'package require Itcl'

### DISCUSSION THREAD

**Discussion by bigfaceworm** [universal-ctags/ctags#3136] [Q&A] [answered]
Tried using the new tags package, but it's not finding nearly as much in my Tcl files.
Looking into the unit tests, and the comments therein, it appears that the new parser(s) for Itcl do not activate unless the file contains 'package require Itcl'.  

How do I force the parser for Itcl (and TclOO) to always be active when parsing a Tcl file?

Our code base uses Itcl in 90% of Tcl files, but due to the flow, doesn't have 'package require Itcl' in but a small handful.

**Comment by masatake** [comment]
>  but due to the flow

Could you tell me the detail of the flow?

ctags needs a hint for detecting whether the current input is Itcl source or non-Itcl tcl source.

**Comment by masatake** [comment]
In the current implementation, ctags detects iTcl in two ways:

* https://github.com/universal-ctags/ctags/blob/f2dc7449db3ed43676e49a52386664132a6c48df/parsers/itcl.c#L270

* https://github.com/universal-ctags/ctags/blob/f2dc7449db3ed43676e49a52386664132a6c48df/parsers/itcl.c#L278

**Reply by bigfaceworm** [reply]
The Tcl file just looks like this:

`itcl::class MyClass {
   public method foo {}
   private method bar {args} {}
}

itcl::body MyClass:foo {} {
}
itcl::body MyClass::bar {} {
}`

There's no 'package require'.

We have 1500 Tcl files, only 250 have 'package require Itcl'.

**Comment by masatake** [comment]
About tcl files having 'package require Itcl', does ctags work fine?

**Reply by bigfaceworm** [reply]
Yes, it appears to work fine, it is just not finding the information about the itcl classes/methods in files that do not have 'package require Itcl'

**Comment by masatake** [comment]
I read my code. As you wrote, itcl.c expects 'package require'. Without the statement, itcl.c does nothing.

To use itcl in your code, isn't "package require ITcl" needed?
I wonder why your code works without the statement.

What we have to do is simple, removing:

https://github.com/universal-ctags/ctags/blob/f2dc7449db3ed43676e49a52386664132a6c48df/parsers/itcl.c#L259-L260

However, I need the reason why we have to remove the condition from the view of Tcl language semantics.

**Reply by bigfaceworm** [reply]
Imagine the following

file1.tcl:
`package require Itcl
 source file2.tcl
 source file3.tcl`

file2.tcl:
`itcl::class MyClass { .... }`

file 3.tcl
`itcl::class YourClass { .... }'

etc.
In fact, if you had a single Tcl interpreter in which these files were sourced, you'd only need one 'package require Itcl'

And, even more, you can easily (in Tcl) set up auto loading such that you *never* need 'package require Itcl' in your .tcl files.
(look at documentation for the 'auto_load' command and the 'auto_index' variable).

**Reply by masatake** [reply]
See #3139 .

**Comment by masatake** [comment, marked as answer]
@bigfaceworm, thank you.
I will remove the condition, `itcl->foundITclPackageRequired`.
Please, wait a while.

**Reply by bigfaceworm** [reply]
Many thanks.  If you want to provide that functionality as a command line (forcing Itcl parsing when parsing Tcl), I'm open to that too.

**Reply by masatake** [reply]
>  If you want to provide that functionality as a command line (forcing Itcl parsing when parsing Tcl), I'm open to that too.

I wonder when we want to provide the extended command line interface.

My guessing:

file1.tcl:
```
package require Itcl
namespace import itcl::*
source file2.tcl
```

file2.tcl:
```
class MyClass { .... }
```

In this case, do you expect MyClass is tagged by the ITcl parser?
In the other words, file2.tcl is a valid ITcl program file?

**Reply by bigfaceworm** [reply]
Yes, I'd expect MyClass to be tagged by the ITcl parser.
file2.tcl is a valid ITcl file.  It is not self-contained, but that is immaterial.

**Reply by masatake** [reply]
Thank you. I have one more question which is more popular、 ITcl or TclOO?

**Reply by bigfaceworm** [reply]
I don't know the answer to that, it kind of depends on the application.  TclOO is much newer, and ITcl has been used in a lot of legacy code (that's my situation).  ITcl also has been used in building graphical widgets, which are fairly popular.  So I'd guess ITcl, but I don't know for sure.

**Reply by masatake** [reply]
Thank you. I will introduce a command-line option specifying a sub parser (TclOO or ITcl) as `class` keyword handler.
